xtask/src/flash_image.rs

```diff
@@ -0,0 +1,504 @@
+            image_count: image_info.len() as u16,
+        };
+
+        let payload = FlashImagePayload {
+            image_info,
+            images: images.to_owned(),
+        };
+
+        let checksum = FlashImageChecksum::new(&header, &payload);
+
+        Self {
+            header,
+            checksum,
+            payload,
+        }
+    }
+
+    fn align_to_4_bytes(data: &mut Vec<u8>) {
+        let padding = (4 - (data.len() % 4)) % 4;
+        data.extend(vec![0; padding]);
+    }
+
+    pub fn write_to_file(&self, filename: &str) -> io::Result<()> {
+        let mut file = File::create(filename)?;
+
+        // Write header
+        file.write_all(&self.header.magic_number.to_le_bytes())?;
+        file.write_all(&self.header.header_version.to_le_bytes())?;
+        file.write_all(&self.header.image_count.to_le_bytes())?;
+
+        // Write checksums
+        file.write_all(&self.checksum.header.to_le_bytes())?;
+        file.write_all(&self.checksum.payload.to_le_bytes())?;
+
+        // Write image info
+        for info in &self.payload.image_info {
+            file.write_all(&info.identifier.to_le_bytes())?;
+            file.write_all(&info.image_offset.to_le_bytes())?;
+            file.write_all(&info.size.to_le_bytes())?;
+        }
+
+        // Write images
+        for image in &self.payload.images {
+            file.write_all(&image.data)?;
+        }
+
+        Ok(())
+    }
+
+    pub fn verify_flash_image(image: &[u8]) -> Result<(), DynError> {
+        // Parse and verify header
+        let magic_number = u32::from_le_bytes(image[0..4].try_into().unwrap());
+        let header_version = u16::from_le_bytes(image[4..6].try_into().unwrap());
+        let image_count = u16::from_le_bytes(image[6..8].try_into().unwrap());
+
+        if magic_number != FLASH_IMAGE_MAGIC_NUMBER {
+            // Return error
+            return Err("Invalid header: incorrect magic number or header version.")?;
+        }
+
+        if header_version != HEADER_VERSION {
+            return Err("Unsupported header version")?;
+        }
+
+        if image_count < 3 {
+            return Err("Expected at least 3 images")?;
+        }
+
+        // Parse and verify checksums
+        let header_checksum = u32::from_le_bytes(image[8..12].try_into().unwrap());
+        let payload_checksum = u32::from_le_bytes(image[12..16].try_into().unwrap());
+        let calculated_header_checksum = FlashImageChecksum::calculate_checksum(&image[0..8]);
+        let calculated_payload_checksum = FlashImageChecksum::calculate_checksum(&image[16..]);
+
+        if header_checksum != calculated_header_checksum {
+            return Err("Header checksum mismatch.")?;
+        }
+
+        if payload_checksum != calculated_payload_checksum {
+            return Err("Payload checksum mismatch.")?;
+        }
+
+        // Parse and verify image info and data
+        let mut offset = 16; // Start after header and checksums
```

Comment:
Rather than having a `mut` here, could you just calculate `let offset = 16 + i * 12` in the `for`-loop?

Reply:
ack